File: scripts/cpp/total_resource.cpp
#include <ros/ros.h>
#include <std_msgs/Float32MultiArray.h>
#include <nvml.h>
#include <sys/sysinfo.h>
#include <fstream>
#include <vector>
#include <string>
#include <iostream>
#include <numeric>

class SystemStatusMonitor {
public:
    SystemStatusMonitor() : rate(1.0), gpu_initialized(false) {
        total_resource_pub = nh.advertise<std_msgs::Float32MultiArray>("/total_resource", 100);

        // Initialize NVML for GPU monitoring
        nvmlReturn_t result = nvmlInit();
        if (result == NVML_SUCCESS) {
            nvmlDeviceGetHandleByIndex(0, &gpu_handle);
            char gpu_name[NVML_DEVICE_NAME_BUFFER_SIZE];
            nvmlDeviceGetName(gpu_handle, gpu_name, NVML_DEVICE_NAME_BUFFER_SIZE);
            gpu_initialized = true;
            ROS_INFO("NVML initialized GPU: %s", gpu_name);
        } else {
            ROS_WARN("GPU initialization failed: %s", nvmlErrorString(result));
        }
    }

    ~SystemStatusMonitor() {
        if (gpu_initialized) {
            nvmlShutdown();
            ROS_INFO("GPU resources released.");
        }
    }

    void run() {
        while (ros::ok()) {
            std_msgs::Float32MultiArray msg;

            // Collect system information
            std::vector<float> cpu_info = getCpuUsage();
            std::vector<float> mem_info = getMemoryUsage();
            std::vector<float> gpu_info = getGpuInfo();

            // Populate message
            msg.data.insert(msg.data.end(), cpu_info.begin(), cpu_info.end());
            msg.data.insert(msg.data.end(), mem_info.begin(), mem_info.end());
            msg.data.insert(msg.data.end(), gpu_info.begin(), gpu_info.end());

            total_resource_pub.publish(msg);

            // Log system information
            ROS_INFO("CPU: %.2f%%, Temp: %.2fC, Freq: %.2f GHz", cpu_info[0], cpu_info[2], cpu_info[1]);
            ROS_INFO("Mem: %.2f/%.2f MB (%.2f%%)", mem_info[0], mem_info[1], mem_info[2]);

            if (gpu_initialized) {
                ROS_INFO("GPU Usage: %.2f%%, Mem: %.2f/%.2f MB (%.2f%%), Temp: %.2fC",
                         gpu_info[0], gpu_info[1], gpu_info[2], gpu_info[3], gpu_info[4]);
            }

            rate.sleep();
        }
    }

private:
    ros::NodeHandle nh;
    ros::Publisher total_resource_pub;
    ros::Rate rate;

    nvmlDevice_t gpu_handle;
    bool gpu_initialized;

    std::vector<float> getCpuUsage() {
        static std::ifstream cpu_stat_file("/proc/stat");
        static std::vector<long> prev_cpu_times(4, 0);

        std::string line;
        std::getline(cpu_stat_file, line);
        cpu_stat_file.clear();
        cpu_stat_file.seekg(0, std::ios::beg);

        std::istringstream iss(line);
        std::string cpu_label;
        iss >> cpu_label;

        std::vector<long> cpu_times(4, 0);
        for (int i = 0; i < 4; ++i) {
            iss >> cpu_times[i];
        }

        long total_diff = std::accumulate(cpu_times.begin(), cpu_times.end(), 0L) -
                          std::accumulate(prev_cpu_times.begin(), prev_cpu_times.end(), 0L);
        long idle_diff = cpu_times[3] - prev_cpu_times[3];

        prev_cpu_times = cpu_times;

        float usage = (total_diff - idle_diff) * 100.0 / total_diff;
        return {usage, 0.0, 0.0};  // Frequency and temperature are placeholders for simplicity
    }

    std::vector<float> getMemoryUsage() {
        struct sysinfo info;
        sysinfo(&info);

        float total_mem = info.totalram / (1024.0 * 1024.0);
        float used_mem = (info.totalram - info.freeram) / (1024.0 * 1024.0);
        float mem_usage = (used_mem / total_mem) * 100.0;

        return {used_mem, total_mem, mem_usage};
    }

    std::vector<float> getGpuInfo() {
        if (!gpu_initialized) {
            return {0, 0, 0, 0, 0};
        }

        nvmlUtilization_t utilization;
        nvmlMemory_t memory;
        unsigned int temperature;

        nvmlDeviceGetUtilizationRates(gpu_handle, &utilization);
        nvmlDeviceGetMemoryInfo(gpu_handle, &memory);
        nvmlDeviceGetTemperature(gpu_handle, NVML_TEMPERATURE_GPU, &temperature);

        float gpu_usage = utilization.gpu;
        float gpu_mem_used = memory.used / (1024.0 * 1024.0);
        float gpu_mem_total = memory.total / (1024.0 * 1024.0);
        float gpu_mem_usage = (gpu_mem_used / gpu_mem_total) * 100.0;
        float gpu_temp = temperature;

        return {gpu_usage, gpu_mem_used, gpu_mem_total, gpu_mem_usage, gpu_temp};
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "total_resource");

    SystemStatusMonitor monitor;

    try {
        monitor.run();
    } catch (const ros::Exception& e) {
        ROS_ERROR("ROS Exception: %s", e.what());
    }

    return 0;
}

File: scripts/total_resource.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import rospy
from std_msgs.msg import Float32MultiArray
import psutil
from pynvml import nvmlInit, nvmlDeviceGetHandleByIndex, nvmlDeviceGetUtilizationRates, nvmlDeviceGetMemoryInfo, nvmlDeviceGetTemperature, nvmlDeviceGetName, nvmlShutdown, NVML_TEMPERATURE_GPU, NVMLError, nvmlDeviceGetCount


class SystemStatusMonitor:
    def __init__(self):
        """Initialize the System Status Monitor."""
        # ROS Publisher 초기화
        self.total_resource_pub = rospy.Publisher('/total_resource', Float32MultiArray, queue_size=100)
        self.rate = rospy.Rate(1)  # 1Hz

        # GPU 초기화
        self.gpu_initialized = False
        try:
            nvmlInit()
            self.gpu_handle = nvmlDeviceGetHandleByIndex(0)
            self.gpu_initialized = True
            rospy.loginfo(f"NVML initialized GPU: {nvmlDeviceGetName(self.gpu_handle)}")
        except NVMLError as e:
            rospy.logwarn(f"GPU initialization failed: {e}")

    def get_cpu_usage(self):
        '''
        psutil.cpu_times_percent return:
        (user, nice, system, idle, iowait, irq, softirq, steal, guset, guset_nice)
        '''
        cpu_usage_total_percent = psutil.cpu_percent(interval=1)
        cpu_usage = psutil.cpu_times_percent(interval=1) #Linux
        
        cpu_freq = psutil.cpu_freq().current / 1000 if psutil.cpu_freq() else -1
        try:
            temps = psutil.sensors_temperatures()
            if 'coretemp' in temps: #x86 cpu
                core_temps = temps['coretemp']
                cpu_temp = sum([t.current for t in core_temps]) / len(core_temps)
            elif 'cpu-thermal' in temps: #arm cpu
                cpu_temp = temps['cpu-thermal'][0].current
            else:
                cpu_temp = -1
        except Exception as e:
            cpu_temp = -1
        return [cpu_usage, cpu_freq, cpu_temp, cpu_usage_total_percent]

    def get_gpu_info(self):
        """return gpu info"""
        if not self.gpu_initialized:
            return []
        
        gpu_info_lst = []
        
        try:
            gpu_name = nvmlDeviceGetName(self.gpu_handle)
            gpu_usage = nvmlDeviceGetUtilizationRates(self.gpu_handle).gpu
            gpu_memory = nvmlDeviceGetMemoryInfo(self.gpu_handle)
            gpu_temp = nvmlDeviceGetTemperature(self.gpu_handle, NVML_TEMPERATURE_GPU)
            
            gpu_info_lst.append([
                0, #id
                gpu_name, #model
                gpu_usage, #gpu usage percent
                gpu_memory.used / (1024**2), #gpu used mem (mb)
                gpu_memory.total / (1024**2), #gpu total mem (mb)
                (gpu_memory.used / gpu_memory.total) * 100 if gpu_memory.total > 0 else -1, #gpu used mem (%)
                gpu_temp #gpu temperature
            ])
                
        except Exception as e:
            rospy.logwarn(f"Failed to get GPU usage: {e}")
            
        return gpu_info_lst
    
    def get_memory_usage(self):
        """Get system memory usage."""
        memory = psutil.virtual_memory()
        return [
            memory.used / (1024**2), #ram used (mb)
            memory.total / (1024**2), #ram total size (mb)
            memory.percent, #ram used (%)
        ]


    def run(self):
        """Main loop to continuously monitor and publish system status."""
        while not rospy.is_shutdown():
            cpu_info = self.get_cpu_usage()
            gpu_info = self.get_gpu_info()
            mem_info = self.get_memory_usage()
            
            msg = Float32MultiArray()
            
            msg.data = cpu_info + mem_info
            
            for gpu in gpu_info:
                msg.data.extend([
                    gpu[0], #gpu id
                    gpu[2], #gpu usage %
                    gpu[3], #gpu used mem mb
                    gpu[4], #gpu total mem mb
                    gpu[5], #gpu mem usage %
                    gpu[6], #gpu temp
                ])
                
            self.total_resource_pub.publish(msg)
            
            rospy.loginfo(f'CPU: user {cpu_info[0][0]:.2f}, system {cpu_info[0][2]}, idle {cpu_info[0][3]}, iowait {cpu_info[0][4]}'
                          f'| Freq: {cpu_info[1]:.2f}GHz | Temp: {cpu_info[2]:.2f}C')
            
            rospy.loginfo(f'Mem: {mem_info[0]:.2f}/{mem_info[1]:.2f} MB ({mem_info[2]:.2f}%)')

            # rospy.loginfo(f'GPU {gpu[0]:.2f}: {gpu[1]} | Usage: {gpu[2]:.2f}% | '
            #               f'GPU Mem: {gpu[3]:.2f}/{gpu[4]:.2f} MB {gpu[5]:.2f}% | Temp: {gpu[6]:.2f}°C')
            rospy.loginfo(f'GPU: {gpu[2]:.2f}% | GPU Mem: {gpu[3]:.2f}/{gpu[4]:.2f} MB ({gpu[5]:.2f}%) | Temp: {gpu[6]:.2f}C')
            
            self.rate.sleep()
            
    def shutdown(self):
        """Clean up resources on shutdown."""
        if self.gpu_initialized:
            nvmlShutdown()
            rospy.loginfo("GPU resources released.")


def main():
    rospy.init_node('total_resouce', anonymous=True)
    monitor = SystemStatusMonitor()

    try:
        monitor.run()
    except rospy.ROSInterruptException:
        pass
    finally:
        monitor.shutdown()


if __name__ == '__main__':
    main()